common/persistence/sql/sqlVisibilityStore.go

```diff
@@ -0,0 +1,351 @@
+				time.Unix(0, request.EarliestStartTime),
+				readLevel.Time,
+				readLevel.RunID,
+				readLevel.Time,
+				request.PageSize)
+		})
+}
+
+func (s *sqlVisibilityStore) ListClosedWorkflowExecutionsByStatus(request *p.ListClosedWorkflowExecutionsByStatusRequest) (*p.ListWorkflowExecutionsResponse, error) {
+	return s.listWorkflowExecutions("ListClosedWorkflowExecutionsByStatus", request.NextPageToken, request.EarliestStartTime, request.LatestStartTime,
+		func(readLevel *visibilityPageToken, rows *[]executionVisibilityRow) error {
+			return s.db.Select(rows,
+				templateGetClosedWorkflowExecutionsByStatus,
+				request.Status,
+				request.DomainUUID,
+				time.Unix(0, request.EarliestStartTime),
+				readLevel.Time,
+				readLevel.RunID,
+				readLevel.Time,
+				request.PageSize)
+		})
+}
+
+func (s *sqlVisibilityStore) GetClosedWorkflowExecution(request *p.GetClosedWorkflowExecutionRequest) (*p.GetClosedWorkflowExecutionResponse, error) {
+	var row executionVisibilityRow
+	execution := request.Execution
+	if err := s.db.Get(&row, templateGetClosedWorkflowExecution, request.DomainUUID, execution.RunId); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, &workflow.EntityNotExistsError{
+				Message: fmt.Sprintf("Workflow execution not found.  WorkflowId: %v, RunId: %v",
+					execution.GetWorkflowId(), execution.GetRunId()),
+			}
+		}
+		return nil, &workflow.InternalServiceError{
+			Message: fmt.Sprintf("GetClosedWorkflowExecution operation failed. Select failed: %v", err),
+		}
+	}
+	row.DomainID = request.DomainUUID
+	row.RunID = execution.GetRunId()
+	row.WorkflowID = execution.GetWorkflowId()
+	return &p.GetClosedWorkflowExecutionResponse{Execution: rowToInfo(row)}, nil
+}
+
+func rowToInfo(row executionVisibilityRow) *workflow.WorkflowExecutionInfo {
+	info := &workflow.WorkflowExecutionInfo{
+		Execution: &workflow.WorkflowExecution{
+			WorkflowId: common.StringPtr(row.WorkflowID),
+			RunId:      common.StringPtr(row.RunID),
+		},
+		Type: &workflow.WorkflowType{Name: common.StringPtr(row.WorkflowTypeName)},
+	}
+	if row.CloseStatus != nil {
+		status := workflow.WorkflowExecutionCloseStatus(*row.CloseStatus)
+		info.CloseStatus = &status
+		closeTime := row.CloseTime.Unix()
+		info.CloseTime = &closeTime
+		info.HistoryLength = row.HistoryLength
+	}
+	return info
+}
+
+func (s *sqlVisibilityStore) listWorkflowExecutions(opName string, pageToken []byte, earliestTime int64, latestTime int64, selectOp func(readLevel *visibilityPageToken, rows *[]executionVisibilityRow) error) (*p.ListWorkflowExecutionsResponse, error) {
+	var rows []executionVisibilityRow
+	var readLevel *visibilityPageToken
+	var err error
+	if len(pageToken) > 0 {
+		readLevel, err = s.deserializePageToken(pageToken)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		readLevel = &visibilityPageToken{Time: time.Unix(0, latestTime), RunID: ""}
+	}
+	err = selectOp(readLevel, &rows)
+	if err != nil {
+		return nil, &workflow.InternalServiceError{
```

Comment:
This error is only returned from query a single row which is not the case here.

Reply:
> This error is only returned from query a single row which is not the case here
- I am not sure if there is a standard across different SQL driver implementations that honors this rule
- Its still defensive to add a check for ErrNoRows and return an empty list (instead of error) to the caller

Reply:
This is not coming from an sql driver, but from sqlx library: https://github.com/jmoiron/sqlx/blob/master/sqlx.go#L209